---
title: "PCA"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

```{r libs, message = FALSE, warning = FALSE}
library(dplyr) # %in%
library(plotfunctions)
library(RColorBrewer)
library(gridExtra)
```

#### Set working directory (full path of the 'scRNAseq_Pipeline' folder).

```{r wd}
setwd("C:/Users/Vlatka/Documents/scRNAseq_Pipeline")
```

#### Import functions

```{r import functions, message = FALSE, warning = FALSE}
# source('Functions/Fun_variableGenes.R')
source('Functions/Fun_prcomp_expressionPlot.R')
source('Functions/Fun_densityPlot.R')
```


### 1 Data import

Import UMI counts. Import gene features.

```{r get data}
counts <- read.csv("./Data/norm_counts.csv", row.names = 1)

# Gene Info
info <- read.csv('./Data/geneInfo_All.csv', header = T, row.names = 1)
```

#### log10(UMI counts)

```{r log expression}
# log10 expression. <1 equaled to 1, resulting in min(log10(x))=0
# For gene expression overlay; doesn't matter if the count is 0 or 1.
logCounts <- counts %>% replace(., . < 1, 1) %>% log10()
```


### 2 PCA

#### Gene selection

``` {r gene selection}
#---------GENE SELECTION-----------
# A) Only keep genes with mean expression above the threshold
exprThreshold <- 0.01 # 9698
data <- counts[rowMeans(counts) > exprThreshold,]

#--- B) Only keep variable genes
# Using 'distance from median' (DM) value as a variance measure
# from Kolodziejczyk et al. 2015 (doi: 10.1016/j.stem.2015.09.011)
# exprThreshold <- 0.01
# perc <- 0.5 # % of genes to keep (ordered by maximum DM)
# geneNames = variableGenes(counts, exprThreshold, perc)
# data <- counts[geneNames,] 
```

#### PCA

```{r pca}
#---------PCA-----------
# expression>0.01 => 14 min (i7-4790 CPU, 16GB, 4 cores/ 8 logical cores)
t0 <- Sys.time()
pca = prcomp(log10(t(data) + 0.001), center = T, scale = T) #log10, standardised
Sys.time() - t0 
```


### 3 Plot gene expression

```{r plot functions, message = FALSE, warning = FALSE}
#---------Plot individual gene-----------
# Simplified plot
plot_pca <- function(dim1, dim2, gene, if_legend, bgcolor = 'white'){
  # par(mfrow = c(1,1), oma = c(0,0,0,0))
  par(mar = c(3,3,1,1))
  plotExpr(dim1, dim2, gene, colPal, steps, psize1, psize2, if_legend, 
           if_panels = F, bgcol = bgcolor)
}

#---------Plot gene set------------------
plot_pcaSet <- function(dim1, dim2, gene_set, thresh, bgcolor = 'white'){
  # Above threshold in at least one cell
  over <- rowSums(counts>thresh) > 0
  set <- info[ ,gene_set] & over
  if (sum(set)==0) return()
  
  # Mean gene expressi on values
  arg = colMeans(logCounts[set, ])
  
  # Plot
  par(mar = c(3,3,1,1))
  plotVector(dim1, dim2, arg, colPal, steps, psize1, psize2, if_legend = F, 
             if_panels = F, bgcol = bgcolor)
  # Title
  title(main = paste0(gene_set,' > ',thresh), cex.main = 0.9)
}
```

```{r plot settings}
#---------Plot settings-----------
# Colour palette
colPal = colorRampPalette(c('gray87', 'whitesmoke', 'gold2', 'orange', 'red'));
steps = 10 #for colour-coding the data

# Outer ring, inner circle
psize1 = 1; psize2 = .75
```

#### Plot gene sets

```{r plots, fig.height = 6, fig.width = 9}
# Genes subselection: more than (threshold) UMIs in at least one cell
th <- 10

# Plot
par(mfrow = c(2,3))
plot_pcaSet(1, 2, 'rib', th)
plot_pcaSet(1, 2, 'agg', th)
plot_pcaSet(1, 2, 'moundUp', th) # Rosengarten 2015
plot_pcaSet(1, 3, 'agg', th)
plot_pcaSet(1, 4, 'agg', th)
```

#### Plot individual genes

```{r, fig.height = 3, fig.width = 6}
par(mfrow = c(1,2))
plot_pca(1, 5, 'pspA', T)
plot_pca(1, 5, 'tps3', T)
```

### 4 Plot cell density

```{r plot cell density, fig.height = 3, fig.width = 12}
#---------Colour palette----------
# Adjust discrete palette length & binNum 
# Reverse & add white as a baseline
mypalette = rev(RColorBrewer::brewer.pal(11, 'RdYlBu')); mypalette =  c('white', mypalette)

#---------Plot----------
grid.arrange(
  plotContour_pca('ndensity', dim_2 = 2, y_lim = c(-40, 30)),
  plotContour_pca('ndensity', dim_2 = 3, y_lim = c(-70, 50)),
  plotContour_pca('ndensity', dim_2 = 4, y_lim = c(-40, 40)),
  plotContour_pca('ndensity', dim_2 = 5, y_lim = c(-100, 35)),
  nrow = 1)
```

